src/01_getting_started/02_why_async.md

```diff
@@ -1,45 +1,121 @@
+large amount of IO-bound tasks, such as servers and databases.
+All else equal, you can have orders of magnitude more tasks than OS threads.
+That said, async Rust can result in larger binary sizes, both due to the state
+machines that are generated from async functions and because each executable
+needs to bundle an async runtime.
+
+### Example: Concurrent downloading
+
+In this example, our goal is to download two web pages concurrently.
+In a typical threaded application, we need to spawn threads
+to achieve concurrency:
 
 ```rust,ignore
 {{#include ../../examples/01_02_why_async/src/lib.rs:get_two_sites}}
 ```
 
-This works fine for many applications—after all, threads were designed
-to do just this: run multiple different tasks at once. However, they also
-come with some limitations. There's a lot of overhead involved in the
-process of switching between different threads and sharing data between
-threads. Even a thread which just sits and does nothing uses up valuable
-system resources. These are the costs that asynchronous code is designed
-to eliminate. We can rewrite the function above using Rust's
-`async`/`.await` notation, which will allow us to run multiple tasks at
-once without creating multiple threads:
+However, downloading a web page is a small task; creating a thread
+for such a small amount of work is quite wasteful. For a larger application, it
+can easily become a bottleneck. In async Rust, we can run these tasks
+concurrently without extra threads:
 
 ```rust,ignore
 {{#include ../../examples/01_02_why_async/src/lib.rs:get_two_sites_async}}
 ```
 
-Overall, asynchronous applications have the potential to be much faster and
-use fewer resources than a corresponding threaded implementation. However,
-there is a cost. Threads are natively supported by the operating system,
-and using them doesn't require any special programming model—any function
-can create a thread, and calling a function that uses threads is usually
-just as easy as calling any normal function. However, asynchronous functions
-require special support from the language or libraries.
-In Rust, `async fn` creates an asynchronous function which returns a `Future`.
-To execute the body of the function, the returned `Future` must be run to
-completion.
-
-It's important to remember that traditional threaded applications can be quite
-effective, and that Rust's small memory footprint and predictability mean that
-you can get far without ever using `async`. The increased complexity of the
-asynchronous programming model isn't always worth it, and it's important to
-consider whether your application would be better served by using a simpler
-threaded model.
+Here, no extra threads are created. Additionally, all function calls are statically
+dispatched, and there are no heap allocations!
+However, we need to write the code to be asynchronous in the first place,
+which this book will help you achieve.
+
+## Other concurrency models
```

Comment:
Is this section header supposed to be here? It doesn't fit well with the content below. A short summary paragraph might be a better ending to this section.

Reply:
Changed to "Custom concurrency models in Rust" to be more clear what it is about. A summary section may be a good idea though..